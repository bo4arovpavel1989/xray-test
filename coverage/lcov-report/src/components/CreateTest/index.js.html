<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/components/CreateTest/index.js</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="..\..\..\prettify.css" />
    <link rel="stylesheet" href="..\..\..\base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(..\..\..\sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="..\..\..\index.html">All files</a> / <a href="index.html">src/components/CreateTest</a> index.js
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>16/16</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">50% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>1/2</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>6/6</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>16/16</span>
      </div>
    </div>
  </div>
  <div class='status-line high'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67</td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import React from 'react'
import { withRouter } from 'react-router-dom'
import { postData } from './../../actions'
&nbsp;
class CreateTest extends React.Component {
  constructor (props) {
    super(props)
&nbsp;
    this.state = {
      err: false,
      loading: false,
      name: ''
    }
&nbsp;
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.postData = props.postData || <span class="branch-1 cbranch-no" title="branch not covered" >postData;</span>
  }
&nbsp;
  handleSubmit (e) {
    e.preventDefault();
&nbsp;
    const { name } = this.state;
&nbsp;
    this.setState({ loading: true });
&nbsp;
    return this.postData('test', { name })
               .then(rep =&gt; window.alert('Успешно сохранено!'))
               .catch(window.alert)
               .finally(() =&gt; this.setState({ loading: false }))
  }
&nbsp;
  handleChange (e) {
    const data = {};
&nbsp;
    data[e.target.id] = e.target.value;
&nbsp;
    this.setState(data);
  }
&nbsp;
  render () {
    const { loading } = this.state;
&nbsp;
    return (
      &lt;div className='container'&gt;
        &lt;div className='formArea'&gt;
          &lt;form onSubmit={this.handleSubmit}&gt;
            &lt;div&gt;
              &lt;span&gt;
                &lt;label&gt;
                  Номер теста: &amp;emsp;&lt;input id='name' onChange={this.handleChange} type='number' min='1'/&gt;
                &lt;/label&gt;
              &lt;/span&gt;
              &lt;span&gt;
                &lt;input disabled= { loading } type='submit' value='Создать'/&gt;
              &lt;/span&gt;
            &lt;/div&gt;
          &lt;/form&gt;
        &lt;/div&gt;
      &lt;/div&gt;
    )
  }
}
&nbsp;
export default withRouter(CreateTest);
export { CreateTest };
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Thu May 30 2019 10:31:51 GMT+0300 (RTZ 2 (зима))
</div>
</div>
<script src="..\..\..\prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="..\..\..\sorter.js"></script>
</body>
</html>
