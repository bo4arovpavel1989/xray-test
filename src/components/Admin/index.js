import React from 'react'
import { Link, withRouter } from 'react-router-dom'
import { getData, deleteData, postData, postFile, downloadFile } from './../../actions'

class Admin extends React.Component {
  constructor (props) {
    super(props)

    this.state = {
      questions: [],
      settings: [],
      tests: [],
      submitting: false,
      err: false
    }

    this.deleteObj = this.deleteObj.bind(this);
    this.refreshTest = this.refreshTest.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.saveDb = this.saveDb.bind(this);
    this.handleDbRecovery = this.handleDbRecovery.bind(this);

    this.getData = this.props.getData || getData;
    this.postData = this.props.postData || postData;
    this.postFile = this.props.postFile || postFile;
    this.deleteData = this.props.deleteData || deleteData;
    this.downloadFile = this.props.downloadFile || downloadFile;
  }

  componentDidMount () {
    this.getSettings();
    this.getTests();
    this.getQuestions();
  }

  getSettings () {
    return this.getData('settings')
               .then(settings => this.setState({ settings }))
               .catch(err => this.setState({ err: true }))
  }

  getTests () {
    return this.getData('tests')
               .then(tests => this.setState({ tests }))
               .catch(err => this.setState({ err: true }))
  }

  getQuestions () {
    return this.getData('questions')
               .then(questions => this.setState({ questions }))
               .catch(err => this.setState({ err: true }))
  }

  getTestData () {
    this.getTests();
    this.getQuestions();
  }

  deleteObj (obj, name) {
    let conf = window.confirm('Вы уверены?');

    if (conf) {
      return this.deleteData(`deleteobj/${obj}/${name}`)
                 .then(rep => this.getTestData())
                 .catch(err => this.setState({ err: true }))
    }
  }

  refreshTest (name) {
    return this.postData('test', { name })
               .then(rep => window.alert('Успешно обновлено!'))
               .catch(window.alert)
  }

  handleChange (e) {
    let { value, type } = e.target;
    const { description } = e.target.dataset;
    const name = e.target.id;
    const { settings } = this.state;

    value = this.handleTypes(value, type);

    settings.forEach((tune, i) => {
      if (name === tune.name) settings[i] = { name, value, type, description }
    })

    this.setState({ settings })
  }

  handleTypes (val, type) {
    const typesMap = {
      number: () => Number(val)
    }

    if (typesMap[type]) return typesMap[type](val)
    else return val
  }

  saveDb () {
    this.setState({ submitting: true })

    return this.getData('savedb')
               .then(() => {
                  // Added timeout, otherwise firefox downloads only 1 file
                  this.downloadFile('download_dump/tests.json', 'tests.json');
                  setTimeout(() =>
                    this.downloadFile('download_dump/questions.json', 'questions.json'),
                  100);
               })
               .catch(window.alert)
               .finally(() => this.setState({ submitting: false }))
  }

  handleDbRecovery (e) {
    e.preventDefault();

    let sure = window.confirm('Вы уверены? Текущая база данных будет перезаписана!')

    if (!sure) return;

    this.setState({ submitting: true })

    const data = this.props.handleFormData || this.handleFormData()

    return this.postFile('loaddb', data)
               .then(() => {
                  window.alert('Успешно восстановлено!');
                  this.getTestData();
               })
               .catch(window.alert)
               .finally(() => this.setState({ submitting: false }))
  }

  handleFormData () {
    const questions = document.querySelector('input[type="file"][name="questions"]');
    const tests = document.querySelector('input[type="file"][name="tests"]');
    const data = new FormData();

    data.append('questions', questions.files[0]);
    data.append('tests', tests.files[0]);

    return data;
  }

  handleSubmit (e) {
    e.preventDefault();
    this.setState({ submitting: true })

    const { settings } = this.state;

    return postData('settings', { settings })
            .then(rep => window.alert('Успешно сохранено!'))
            .catch(window.alert)
            .finally(() => this.setState({ submitting: false }))
  }

  render () {
    const { settings, tests, questions, submitting } = this.state;

    return (
      <div className='container'>
        <div className='newTest'>
          <div className='menuItem'>
            <Link className='menuButton' to='/create/test'>Создать новый тест</Link>
            <Link className='menuButton' to='/create/question'>Создать новый вопрос</Link>
          </div>
          <div className='dbControls'>
            <button disabled={submitting} className='menuButton small saveDb' onClick={this.saveDb}>Сохранить базу</button>
            <form className='loadDb' onSubmit={this.handleDbRecovery}>
              <label> Выберите JSON файл с вопросами &emsp;
                <input type='file' name='questions' accept='.json' required/>
              </label><br/>
              <label> Выберите JSON файл с тестами &emsp;
                <input type='file' name='tests' accept='.json' required/>
              </label><br/>
              <input disabled={submitting} type='submit' value='Восстановить базу'/>
            </form>
          </div>
        </div>
          <h2>Настройки</h2>
          <div className='formArea'>
            <form onSubmit={this.handleSubmit}>
              {
                settings.map(tune => {
                  return <div key={tune._id}>
                    <label>
                      {tune.description}: &emsp;
                      <input
                        id={tune.name}
                        data-description={tune.description}
                        type={tune.type}
                        value={tune.value}
                        onChange={this.handleChange}
                      />
                    </label>
                  </div>
                })
              }
              <div>
                <input disabled={submitting} type='submit' value='Сохранить'/>
              </div>
            </form>
          </div>
        <div>
        </div>
        <div className='oldTest'>
          <h2>Тесты</h2>
          <ul>
            {
              tests.map(test => {
                return <li key={test._id}>{test.name} &emsp;
                <a onClick={() => this.refreshTest(test.name)}>Обновить</a> &emsp;
                <a className='danger' onClick={() => this.deleteObj('Test', test.name)}>Удалить</a>
                </li>
              })
            }
          </ul>
        </div>
        <div className='oldQuestions'>
          <h2>Вопросы</h2>
          <ul>
            {
              questions.map(q => {
                return <li key={q._id}>{q.name} &emsp;
                  <Link to={`/create/question?question=${q.name}`}>Изменить</Link>&emsp;
                  <a className='danger' onClick={() => this.deleteObj('Question', q.name)}>Удалить</a>
                </li>
              })
            }
          </ul>
        </div>
      </div>
    )
  }
}

export default withRouter(Admin)
export { Admin }
