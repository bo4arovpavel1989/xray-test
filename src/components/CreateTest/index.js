import React from 'react'
import { withRouter } from 'react-router-dom'
import { postData } from './../../actions'

class CreateTest extends React.Component {
  constructor (props) {
    super(props)

    this.state = {
      err: false,
      loading: false,
      name: ''
    }

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.postData = props.postData || postData;
  }

  handleSubmit (e) {
    e.preventDefault();

    const { name } = this.state;

    this.setState({ loading: true });

    return this.postData('test', { name })
               .then(rep => window.alert('Успешно сохранено!'))
               .catch(window.alert)
               .finally(() => this.setState({ loading: false }))
  }

  handleChange (e) {
    const data = {};

    data[e.target.id] = e.target.value;

    this.setState(data);
  }

  render () {
    const { loading } = this.state;

    return (
      <div className='container'>
        <div className='formArea'>
          <form onSubmit={this.handleSubmit}>
            <div>
              <span>
                <label>
                  Номер теста: &emsp;<input id='name' onChange={this.handleChange} type='number' min='1'/>
                </label>
              </span>
              <span>
                <input disabled= { loading } type='submit' value='Создать'/>
              </span>
            </div>
          </form>
        </div>
      </div>
    )
  }
}

export default withRouter(CreateTest);
export { CreateTest };
