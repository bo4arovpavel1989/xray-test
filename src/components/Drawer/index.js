class Drawer {
  constructor (canvas) {
    this.canvas = canvas;
    this.ctx = this.canvas.getContext('2d');

    this.zones = [];

    this.x1 = 0;
    this.x2 = 0;
    this.y1 = 0;
    this.y2 = 0;

    this.prevProperties = [0, 0, 0, 0];

    this.clearListeners();
    this.isDrawing = false;

    this.handelMouseUp = this.handelMouseUp.bind(this);
    this.handelMouseMove = this.handelMouseMove.bind(this);
    this.handelMouseDown = this.handelMouseDown.bind(this);
    this.clearZones = this.clearZones.bind(this);
  }

  handelMouseDown (e) {
    this.isDrawing = true;

    this.x1 = e.offsetX;
    this.y1 = e.offsetY;
  }

  handelMouseMove (e) {
    if (this.isDrawing) {
      this.clearPreviousRect();

      this.x2 = e.offsetX;
      this.y2 = e.offsetY;

      const rectProperties = this.getRectProperties();

      this.drawRect(rectProperties);
      this.prevProperties = rectProperties;
    }
  }

  handelMouseUp (e) {
    this.isDrawing = false;

    this.clearPreviousRect();

    this.x2 = e.offsetX;
    this.y2 = e.offsetY;

    const rectProperties = this.getRectProperties();

    this.drawRect(rectProperties);
    this.prevProperties = [0, 0, 0, 0];
    this.saveZones();
  }

  drawRect (rectProperties, color = 'red') {
    this.ctx.strokeStyle = color;
    this.ctx.rect(...rectProperties);
    this.ctx.stroke();
  }

  clearPreviousRect () {
    const [x, y, width, heigth] = this.prevProperties;

    this.ctx.clearRect(x - 1, y - 1, width + 2, heigth + 2);
    this.ctx.beginPath();
  }

  saveZones () {
    const leftUpCorner = this.getLeftUpCorner();
    const rightDownCorner = this.getRightDownCorner();

    this.zones.push([...leftUpCorner, ...rightDownCorner]);
  }

  /**
   * Method draws rectangles got from API
   */
  drawOldZones (rectArray) {
    this.zones = rectArray;

    rectArray.forEach(dimensions => {
      const rectProperties = [dimensions[0], dimensions[1]];
      const width = dimensions[2] - dimensions[0];
      const height = dimensions[3] - dimensions[1];

      rectProperties.push(width);
      rectProperties.push(height);

      this.drawRect(rectProperties);
    })
  }

  getLeftUpCorner () {
    const { x1, x2, y1, y2 } = this;

    return [Math.min(x1, x2), Math.min(y1, y2)]
  }

  getRightDownCorner () {
    const { x1, x2, y1, y2 } = this;

    return [Math.max(x1, x2), Math.max(y1, y2)]
  }

  getRectProperties () {
    const leftUpCorner = this.getLeftUpCorner();
    const rightDownCorner = this.getRightDownCorner();

    let width = rightDownCorner[0] - leftUpCorner[0];
    let height = rightDownCorner[1] - leftUpCorner[1];

    return [...leftUpCorner, width, height];
  }

  start () {
    this.canvas.addEventListener('mousedown', this.handelMouseDown)
    this.canvas.addEventListener('mouseup', this.handelMouseUp)
    this.canvas.addEventListener('mousemove', this.handelMouseMove)
  }

  clearListeners () {
    this.canvas.removeEventListener('mousedown', this.handelMouseDown)
    this.canvas.removeEventListener('mouseup', this.handelMouseUp)
    this.canvas.removeEventListener('mousemove', this.handelMouseMove)
  }

  getZones () {
    return this.zones;
  }

  clearZones () {
    this.zones = [];
    this.ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
    this.ctx.beginPath();
  }

  reset () {
    this.clearListeners();
    this.clearZones();
  }
}

export default Drawer;
