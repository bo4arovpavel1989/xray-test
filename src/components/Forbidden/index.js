import React from 'react'
import { withRouter } from 'react-router-dom'

class Forbidden extends React.Component {
  render () {
    return (
      <div>
        <h1>403 Нет доступа</h1>
        <h3>Залогиньтесь для продолжения</h3>
      </div>
    )
  }
}

export default withRouter(Forbidden)
