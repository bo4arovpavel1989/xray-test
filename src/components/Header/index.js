import React from 'react';
import {Link, withRouter} from 'react-router-dom';
import PropTypes from 'prop-types';
import Login from './Login';
import './Header.sass';

const Header = props=>{
  const {isAdmin, logoff, login} = props;

  return (
    <div className='headerArea'>
      <Link className='header' to='/'>
        <h1>XRay-test</h1>
      </Link>
      {
        isAdmin ?
          <div className='loginArea'>
            <a className='logoff' onClick={logoff}>
              Выйти
            </a>
          </div>
          :
          <Login login={login}/>
        }
    </div>
  )
};

Header.propTypes = {
  isAdmin: PropTypes.bool.isRequired,
  logoff: PropTypes.func.isRequired,
  login: PropTypes.func.isRequired
};

export default withRouter(Header);
