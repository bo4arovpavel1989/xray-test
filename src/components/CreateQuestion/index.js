import React from 'react'
import { withRouter } from 'react-router-dom'
import { postFile, postData, getData } from './../../actions'
import Drawer from '../Drawer'
import './CreateQuestion.sass'

class CreateQuestion extends React.Component {
  constructor () {
    super()

    this.state = {
      loading: false,
      err: false,
      imgPath: '',
      dimensions: { width: 0, height: 0 },
      dangerZones: [],
      name: '',
      isDanger: '1'
    }

    this.handleChange = this.handleChange.bind(this);
    this.uploadPhotos = this.uploadPhotos.bind(this);
    this.clearCanvas = this.clearCanvas.bind(this);
    this.saveQuestion = this.saveQuestion.bind(this);
  }

  componentDidMount () {
    const canvasDraw = '#canvasDrawArea';
    const canvasBackground = '#canvasBackground';
    this.setState({ canvasBackground, canvasDraw })

    this.drawer = new Drawer(document.querySelector(canvasDraw));
    this.checkIfQuestionCreated()
  }

  componentWillUnmount () {
    this.drawer.reset();
  }

  /**
   * Get selector for canvas draw layer
   * @returns {Object} - DOM selector
   */
  getCanvasDraw () {
    return document.querySelector(this.state.canvasDraw)
  }

  /**
   * Get selector for canvas background layer
   * @returns {Object} - DOM selector
   */
  getCanvasBackground () {
    return document.querySelector(this.state.canvasBackground)
  }

  /**
   * Method checks location query
   * if query is defined ( /create/question?question=name )
   * it means user want to edit old question
   * so we fetch question to API
   * @return {Function} - fetch to API for question data
   */
  checkIfQuestionCreated () {
    if (this.props.location.search.length > 0) {
      const query = new URLSearchParams(this.props.location.search);
      const question = query.get('question');

      return this.getQuestion(question)
    }
  }

  /**
   * Method fetches to API for question data
   * if any - renders it
   * @returns {Function} - fetch to API
   */
  getQuestion (question) {
    return getData(`question/${question}`)
            .then(rep => {
              console.log(rep)
              if (rep) {
                this.setState(rep, this.prepareCanvas)
              }
            })
            .catch(err => window.alert(err))
  }

  /**
   * Method handles upload of slide and photo
   * then it gets dimensions of slide and prepares canvas for it
   * @return {void}
   */
  uploadPhotos (e) {
    e.preventDefault();

    const data = this.handleFormData();

    this.setState({ loading: true });
    postFile('preupload', data)
      .then(rep => {
        const { name } = this.state;
        const imgPath = `/images/${name}_slide.${rep.type}`;
        this.drawer.reset();

        this.setState({
          loading: false,
          imgPath: imgPath,
          dimensions: rep
        }, this.prepareCanvas)
      })
      .catch(err => {
        this.setState({ loading: false })
        window.alert(err)
      });
  }

  /**
   * Method gatheres all formData to single object for upload
   * @return {Object} - object to upload
   */
  handleFormData () {
    const slide = document.querySelector('input[name="slide"]');
    const photo = document.querySelector('input[name="photo"]');
    const { name, isDanger } = this.state;
    const data = new FormData();

    data.append('question', name);
    data.append('slide', slide.files[0]);
    if (isDanger === '1') data.append('photo', photo.files[0]);

    return data;
  }

  /**
   * Method handles changes in input and pass it to state
   * @returns {void}
   */
  handleChange (e) {
    const data = {};

    data[e.target.id] = e.target.value;

    this.setState(data);
  }

  /**
   * Method sets canvas width and heigth, add background image
   * and inits drawing dangerzone rectangels
   * @returns {void}
   */
  prepareCanvas () {
    this.setCanvasDimensions();

    const background = new Image();

    background.src = this.state.imgPath;
    background.onload = () => {
      this.setCanvasBackground(background)
      this.drawDangerZone();
    }
  }

  /**
   * Method set canvas width and heigth - both for background and draw layers
   * @returns {void}
   */
  setCanvasDimensions () {
    const { width, height } = this.state.dimensions;

    this.getCanvasDraw().width = width;
    this.getCanvasDraw().height = height;
    this.getCanvasBackground().width = width;
    this.getCanvasBackground().height = height;
  }

  /**
   * Method calls background layer canvas to draw background
   * @returns {Function} - drawImage method of canvas
   */
  setCanvasBackground (bg) {
    return this.getCanvasBackground().getContext('2d').drawImage(bg, 0, 0)
  }

  /**
   * Method draws old rectangles got from API if any
   * and starts to listen to new user draw
   * @returns {Function} - start method of this.draw
   */
  drawDangerZone (c) {
    const { dangerZones } = this.state;

    if (dangerZones.length > 0) this.drawer.drawOldZones(dangerZones);

    return this.drawer.start()
  }

  /**
   * Method calls clearZones mthod of this.drawer
   * to clear canvas draw layer
   */
  clearCanvas () {
    return this.drawer.clearZones()
  }

  /**
   * Method gatheres all question data to single object and post to API
   * @returns {Function} - fetch post API
   */
  saveQuestion () {
    const { imgPath, name, isDanger, dimensions } = this.state;

    const question = {
      name,
      isDanger,
      dangerZones: this.drawer.getZones(),
      imgPath,
      dimensions
     };

     this.setState({ loading: true });

     return postData('savequestion', question)
              .then(rep => {
                window.alert('Успешно сохранено!');
                this.setState({ loading: false })
              })
              .catch(err => {
                window.alert(err);
                this.setState({ loading: false })
              })
  }

  render () {
    const { loading, isDanger, name } = this.state;

    return (
      <div className='container'>
        <div className='formArea'>
          <form onSubmit={this.uploadPhotos}>
            <div>
              <label>
                Вопрос: &emsp;
                <input
                  id='name'
                  onChange={this.handleChange}
                  pattern='[0-9]+_[0-9]+'
                  name='name'
                  type='text'
                  value = {name.length > 0 ? name : ''}
                  placeholder='№ Теста_№ вопроса'
                  required
                />
              </label>
            </div>
            <div>
              <label>
                Загрузите слайд: &emsp;
                <input
                  name='slide'
                  type='file'
                  required
                />
              </label>
            </div>
            <div>
              <label>
                Багаж опасен? &emsp;
                <select
                  id='isDanger'
                  onChange={this.handleChange}
                  value={isDanger === '1' ? '1' : '0'}
                >
                  <option value='1'>Да</option>
                  <option value='0'>нет</option>
                </select>
              </label>
            </div>
            {
              isDanger === '1' ?
              <div>
                <label>
                  Загрузите фото: &emsp;
                  <input
                    name='photo'
                    type='file'
                    required
                  />
                </label>
              </div> :
              <span></span>
            }
            <div>
              <input disabled={loading} type='submit' value='Продолжить'/>
            </div>
          </form>
        </div>
        <div>
          <button onClick={this.clearCanvas}>Очистить</button>
          <button disabled={loading} onClick={this.saveQuestion}>Сохранить</button>
          <div className='canvasArea'>
            <canvas id="canvasBackground"></canvas>
            <canvas id="canvasDrawArea"></canvas>
          </div>
        </div>
      </div>
    )
  }
}

export default withRouter(CreateQuestion)
