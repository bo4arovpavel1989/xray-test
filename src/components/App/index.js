import React from 'react'
import Header from '../Header'
import Menu from '../Menu'
import Test from '../Test'
import Admin from '../Admin'
import Forbidden from '../Forbidden'
import CreateTest from '../CreateTest'
import CreateQuestion from '../CreateQuestion'
import { BrowserRouter, Route, Switch } from 'react-router-dom'
import '../../main.sass'
import { postData } from '../../actions.js'
import { setToken } from '../../helpers'

class App extends React.Component {
  constructor () {
    super()

    this.state = {
      isAdmin: false,
      authFail: false,
      isLogging: false,
      token: undefined
    }

    this.login = this.login.bind(this)
    this.logoff = this.logoff.bind(this)
    this.handleAuth = this.handleAuth.bind(this)
  }

  login (e) {
    e.preventDefault()

    const data = {
      login: e.target.login.value,
      password: e.target.password.value
    }

    this.setState({ isLogging: true })

    postData('login', data)
      .then(this.handleAuth)
      .catch(console.log)
      .finally(() => this.setState({ isLogging: false }))
  }

  handleAuth (rep) {
    if (rep.auth) setToken(rep.token)
    this.setState({ isAdmin: rep.auth, authFail: !rep.auth, token: rep.token })
  }

  logoff () {
    this.setState({
      isAdmin: false,
      token: undefined
     })

     setToken(null)
  }

  render () {
    const { isAdmin, authFail, isLogging } = this.state

    return (
      <BrowserRouter>
        <div>
          <Header
           isAdmin={isAdmin}
           authFail={authFail}
           isLogging = {isLogging}
           login={this.login}
           logoff={this.logoff}
           />
          <Switch>
            <Route path='/' render={Menu} exact/>
            <Route path='/test' render={Test}/>
            <Route path='/admin' render={isAdmin ? Admin : Forbidden}/>
            <Route path='/create/test' render={isAdmin ? CreateTest : Forbidden}/>
            <Route path='/create/question' render={isAdmin ? CreateQuestion : Forbidden}/>
          </Switch>
        </div>
      </BrowserRouter>
    )
  }
}

export default App;
