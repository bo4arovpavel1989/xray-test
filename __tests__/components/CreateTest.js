import React from 'react';
import { CreateTest } from '../../src/components/CreateTest';
import Enzyme, { mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

describe('CreateTest component', () => {
  window.alert = jest.fn();
  const props = {
    postData: jest.fn(() => new Promise((res, rej) => res({ success: true })))
  };
  const initialState = {
    err: false,
    loading: false,
    name: ''
  }
  const component = mount(<CreateTest { ...props }/>);

  it('should render correct', () => {
      expect(shallowToJson(component)).toMatchSnapshot()
  })

  describe('should handle input change', () => {
    beforeAll(() => {
      component.find('#name').simulate('change', {
        target: {
          value: '5',
          id: 'name'
        }
      })
    })

    afterAll(() => {
      component.setState(initialState)
    })

    it('should change state', () => {
      expect(component.state().name).toEqual('5')
    })
  })

  describe('should handle form submit', () => {
    const name = '5';

    beforeAll(() => {
      component.setState({ name }, () => {
        component.find('.formArea form').simulate('submit')
      })
    })

    afterAll(() => {
      component.setState(initialState)
    })

    it('should call postData', () => {
      expect(props.postData).toHaveBeenCalledWith('test', { name })
    })

    it ('should alert success', () => {
      expect(window.alert).toBeCalledWith('Успешно сохранено!')
    })
  })
})
