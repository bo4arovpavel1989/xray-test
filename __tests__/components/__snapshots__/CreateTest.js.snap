// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateTest component should render correct 1`] = `
<CreateTest
  postData={[MockFunction]}
>
  <div
    className="container"
  >
    <div
      className="formArea"
    >
      <form
        onSubmit={[Function]}
      >
        <div>
          <span>
            <label>
              Номер теста:  
              <input
                id="name"
                min="1"
                onChange={[Function]}
                type="number"
              />
            </label>
          </span>
          <span>
            <input
              disabled={false}
              type="submit"
              value="Создать"
            />
          </span>
        </div>
      </form>
    </div>
  </div>
</CreateTest>
`;
